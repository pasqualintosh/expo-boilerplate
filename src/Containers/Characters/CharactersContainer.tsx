import React, { useEffect } from 'react'
import { useLazyFetchCharactersQuery } from '@/Services/modules/rick'
import { useTheme } from '@/Hooks'
import { SafeAreaView } from 'react-native-safe-area-context'
import { FlatList, ListRenderItem, StyleSheet, View } from 'react-native'
import { Result } from '@/Services/modules/rick/fetchCharacters'
import { CharacterItem, Loading, Separator } from '@/Components'

const CharactersContainer = () => {
  const [fetchCharacters, { data, isLoading }] = useLazyFetchCharactersQuery()
  const { Layout, Colors, Gutters } = useTheme()

  useEffect(() => {
    fetchCharacters()
  })

  if (isLoading) {
    return <Loading />
  }

  const renderItem: ListRenderItem<Result> | null | undefined = ({ item }) => (
    <CharacterItem name={item.name} uri={item.image} />
  )

  return (
    <SafeAreaView
      style={[
        Layout.fill,
        Layout.colHCenter,
        Gutters.regularHPadding,
        Gutters.regularVPadding,
        { backgroundColor: Colors.white },
      ]}
    >
      <FlatList
        data={data?.results}
        renderItem={renderItem}
        testID={'charactersList'}
        ItemSeparatorComponent={() => <Separator />}
      />
    </SafeAreaView>
  )
}

const styles = StyleSheet.create({})

export default CharactersContainer
