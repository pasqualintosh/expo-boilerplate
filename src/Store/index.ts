import AsyncStorage from '@react-native-async-storage/async-storage'
import { configureStore, combineReducers } from '@reduxjs/toolkit'
import { setupListeners } from '@reduxjs/toolkit/query'
import {
  persistReducer,
  persistStore,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist'

import { api } from '@/Services/api'
import theme from './Theme'

const reducers = combineReducers({
  theme,
  api: api.reducer,
})

const persistConfig = {
  key: 'root',
  storage: AsyncStorage,
  whitelist: ['theme'],
}

const persistedReducer = persistReducer(persistConfig, reducers)

const store = configureStore({
  reducer: persistedReducer,
  middleware: getDefaultMiddleware => {
    const middlewares = getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }).concat(api.middleware)

    return middlewares
  },
})

const persistor = persistStore(store)

setupListeners(store.dispatch)

export { store, persistor }
