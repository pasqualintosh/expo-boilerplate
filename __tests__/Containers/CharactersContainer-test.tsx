import React from 'react'
import { render, cleanup, act, waitFor } from '@testing-library/react-native'
import { PersistGate } from 'redux-persist/integration/react'
import { Provider } from 'react-redux'
import { store } from '@/Store'
import CharactersContainer from '@/Containers/Characters/CharactersContainer'

jest.mock('@/Services/modules/rick', () => {
  return {
    __esModule: true,
    useLazyFetchCharactersQuery: jest.fn(() => [
      function fetchCharacters() {
        return jest.fn(() => ({
          info: {
            count: 826,
            pages: 42,
            next: 'https://rickandmortyapi.com/api/character/?page=2',
            prev: null,
          },
          results: [
            {
              id: 1,
              name: 'Rick Sanchez',
              status: 'Alive',
              species: 'Human',
              type: '',
              gender: 'Male',
              origin: {
                name: 'Earth',
                url: 'https://rickandmortyapi.com/api/location/1',
              },
              location: {
                name: 'Earth',
                url: 'https://rickandmortyapi.com/api/location/20',
              },
              image: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg',
              episode: [
                'https://rickandmortyapi.com/api/episode/1',
                'https://rickandmortyapi.com/api/episode/2',
              ],
              url: 'https://rickandmortyapi.com/api/character/1',
              created: '2017-11-04T18:48:46.250Z',
            },
          ],
        }))
      },
      {
        data: {
          info: {
            count: 826,
            pages: 42,
            next: 'https://rickandmortyapi.com/api/character/?page=2',
            prev: null,
          },
          results: [
            {
              id: 1,
              name: 'Rick Sanchez',
              status: 'Alive',
              species: 'Human',
              type: '',
              gender: 'Male',
              origin: {
                name: 'Earth',
                url: 'https://rickandmortyapi.com/api/location/1',
              },
              location: {
                name: 'Earth',
                url: 'https://rickandmortyapi.com/api/location/20',
              },
              image: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg',
              episode: [
                'https://rickandmortyapi.com/api/episode/1',
                'https://rickandmortyapi.com/api/episode/2',
                // ...
              ],
              url: 'https://rickandmortyapi.com/api/character/1',
              created: '2017-11-04T18:48:46.250Z',
            },
          ],
        },
        isSuccess: true,
        isLoading: false,
        error: false,
      },
      { lastArg: Symbol() },
    ]),
  }
})

describe('CharactersContainer', () => {
  beforeEach(() => {})
  afterEach(() => cleanup())

  it('match the snapshot', async () => {
    await act(async () => {
      const page = await waitFor(() =>
        render(
          <Provider store={store}>
            <CharactersContainer />
          </Provider>,
          {
            createNodeMock: element => {
              if (element.type === PersistGate) {
                return element.props.children
              }
              return null
            },
          },
        ),
      )

      return new Promise(resolve => {
        setTimeout(() => {
          expect(page).toMatchSnapshot()
          resolve()
        }, 3000)
      })
    })
  })

  it('has charactersList', async () => {
    await act(async () => {
      const page = await waitFor(() =>
        render(
          <Provider store={store}>
            <CharactersContainer />
          </Provider>,
          {
            createNodeMock: element => {
              if (element.type === PersistGate) {
                return element.props.children
              }
              return null
            },
          },
        ),
      )

      return new Promise(resolve => {
        setTimeout(() => {
          expect(page.getAllByTestId('charactersList')).not.toBeNull()
          resolve()
        }, 3000)
      })
    })
  })
})
