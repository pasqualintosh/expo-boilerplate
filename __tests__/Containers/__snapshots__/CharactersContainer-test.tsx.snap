// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharactersContainer match the snapshot 1`] = `
<RNCSafeAreaView
  edges={
    [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
  style={
    [
      {
        "flex": 1,
      },
      {
        "flexDirection": "column",
        "justifyContent": "center",
      },
      {
        "paddingHorizontal": 15,
      },
      {
        "paddingVertical": 15,
      },
      {
        "backgroundColor": "#ffffff",
      },
    ]
  }
>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    data={
      [
        {
          "created": "2017-11-04T18:48:46.250Z",
          "episode": [
            "https://rickandmortyapi.com/api/episode/1",
            "https://rickandmortyapi.com/api/episode/2",
          ],
          "gender": "Male",
          "id": 1,
          "image": "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
          "location": {
            "name": "Earth",
            "url": "https://rickandmortyapi.com/api/location/20",
          },
          "name": "Rick Sanchez",
          "origin": {
            "name": "Earth",
            "url": "https://rickandmortyapi.com/api/location/1",
          },
          "species": "Human",
          "status": "Alive",
          "type": "",
          "url": "https://rickandmortyapi.com/api/character/1",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    testID="charactersList"
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            [
              {
                "flexDirection": "row",
                "justifyContent": "center",
              },
              {
                "flexGrow": 1,
                "justifyContent": "space-between",
              },
            ]
          }
        >
          <Text>
            Rick Sanchez
          </Text>
          <View>
            <Image
              source={
                {
                  "uri": "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
                }
              }
              style={
                {
                  "height": 66,
                  "width": 100,
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RNCSafeAreaView>
`;
